cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -1227,60 +1228,68 @@ CacheAllocator<CacheTrait>::findEviction(PoolId pid, ClassId cid) {
     ++searchTries;
     (*stats_.evictionAttempts)[pid][cid].inc();
 
-    Item* candidate = itr.get();
+    Item* toRecycle = itr.get();
+
+    Item* candidate =
+        toRecycle->isChainedItem()
+            ? &toRecycle->asChainedItem().getParentItem(compressor_)
+            : toRecycle;
+
+    // make sure no other thead is evicting the item
+    if (candidate->getRefCount() != 0 || !candidate->markExclusive()) {
+      ++itr;
+      continue;
+    }
+
     // for chained items, the ownership of the parent can change. We try to
     // evict what we think as parent and see if the eviction of parent
     // recycles the child we intend to.
-    auto toReleaseHandle =
-        itr->isChainedItem()
-            ? advanceIteratorAndTryEvictChainedItem(itr)
-            : advanceIteratorAndTryEvictRegularItem(mmContainer, itr);
+    {
+      auto toReleaseHandle =
+          itr->isChainedItem()
+              ? advanceIteratorAndTryEvictChainedItem(itr)
+              : advanceIteratorAndTryEvictRegularItem(mmContainer, itr);
+      // destroy toReleseHandle. The item won't be release to allocator
+      // since we marked it as exclusive.
+    }
 
-    if (toReleaseHandle) {
-      if (toReleaseHandle->hasChainedItem()) {
+    const auto ref = candidate->unmarkExclusive();
+    if (ref == 0u) {
```

Comment:
I think we still need to check if `toReleaseHandle` is valid. Not because we want to make sure its not being evicted by another thread, but to know whether we did actually removed `candidate` from the `advanceIteratorAndTryEvict*Item` function call.

If an item was not removed, we should skip the logic inside the if-else block and simply unmarkExclusive and move on. If we don't want to return toReleaseHandle because we don't want to increment the refcount, we should find another way to check if the iterator was advanced or not.

Reply:
Does it matter whether **we** did actually remove the `candidate`? If it was some other thread that did this and the ref is `0` after `unmarkExclsuive` we should still be able to recycle the item. `unmarkExclusive` in this case is like `decRef()` . For `ref == 0` we are the last owner but instead of returning the item to allocator we reuse it.

I even belive that we **have** to recycle the item (or release it back to allocator) in this case. Otherwise we could have a memory leak.

Is my thinking correct or do I miss something?

Reply:
I think you are right. `unmarkExclusive` has both admin and access refcount right? if so, you are correct that we would have to recycle as this is the last owner.

I think there is just one small problem now:

C1: removed item and ref == 0 (no problem. The golden path)
C2: removed item and ref != 0 (shouldn't happen. No access ref can be created after removed from AccessContainer)
C3: not removed and ref == 0 (no problem. If happened, should still recycle because another thread had removed the item from Access and MMContainer, but didn't recycle because current thread is holding exclusive bit)
C4: not removed and ref > 0. In this case, we would increment the eviction failure counters again in the else block, while we already incremented them inside `advanceIteratorAndTryEvict`.

I think we just need to fix C4 here.

Reply:
Yes, `ref` will contain both access and admin bits.

You are also right about those 4 cases. Since C2 cannot happen, all the stats are always updated inside `advanceIteratorAndTryEvict*`. So, I just removed the code which updates stats outside of those functions and added an `assert` for C2.